blockindex/sgd.go

```diff
@@ -0,0 +1,235 @@
+// or fitness for purpose and, to the extent permitted by law, all liability for your use of the code is disclaimed.
+// This source code is governed by Apache License 2.0 that can be found in the LICENSE file.
+
+package blockindex
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/iotexproject/go-pkgs/cache"
+	"github.com/iotexproject/go-pkgs/hash"
+	"github.com/iotexproject/iotex-address/address"
+	"github.com/iotexproject/iotex-proto/golang/iotextypes"
+	"google.golang.org/protobuf/proto"
+
+	"github.com/iotexproject/iotex-core/action"
+	"github.com/iotexproject/iotex-core/blockchain/block"
+	"github.com/iotexproject/iotex-core/blockchain/blockdao"
+	"github.com/iotexproject/iotex-core/blockindex/indexpb"
+	"github.com/iotexproject/iotex-core/db"
+	"github.com/iotexproject/iotex-core/db/batch"
+	"github.com/iotexproject/iotex-core/pkg/util/byteutil"
+)
+
+const (
+	_sgdBucket                = "sgd"
+	_sgdToHeightNS            = "hh"
+	_sgdActCreate  sgdActType = iota
+	_sgdActCall
+)
+
+var _sgdCurrentHeight = []byte("currentHeight")
+
+type (
+	// SGDIndexer is the interface for Sharing of Gas-fee with DApps indexer
+	SGDIndexer interface {
+		blockdao.BlockIndexer
+		// CheckContract
+		CheckContract(contract string) (address.Address, uint64, bool)
+		// GetSGDIndex
+		GetSGDIndex(contract string) (*indexpb.SGDIndex, error)
+	}
+
+	sgdRegistry struct {
+		kvStore    db.KVStore
+		kvCache    cache.LRUCache
+		percentage uint64
+	}
+
+	sgdActType int
+
+	sgdAct struct {
+		actType    sgdActType
+		sender     address.Address
+		contract   string
+		createTime time.Time
+	}
+)
+
+// NewSGDRegistry creates a new SGDIndexer
+func NewSGDRegistry(kv db.KVStore, kvCache cache.LRUCache, percentage uint64) SGDIndexer {
+	if kv == nil {
+		panic("nil kvstore")
+	}
+	if percentage > 100 {
+		panic("percentage should be less than 100")
+	}
+	return &sgdRegistry{
+		kvStore:    kv,
+		percentage: percentage,
+		kvCache:    kvCache,
+	}
+}
+
+// Start starts the SGDIndexer
+func (sgd *sgdRegistry) Start(ctx context.Context) error {
+	return sgd.kvStore.Start(ctx)
+}
+
+// Stop stops the SGDIndexer
+func (sgd *sgdRegistry) Stop(ctx context.Context) error {
+	return sgd.kvStore.Stop(ctx)
+}
+
+// Height returns the current height of the SGDIndexer
+func (sgd *sgdRegistry) Height() (uint64, error) {
+	h, err := sgd.kvStore.Get(_sgdToHeightNS, _sgdCurrentHeight)
+	if err != nil {
+		// if db not exist, return 0, nil, after PutBlock, the height will be increased
+		if errors.Is(err, db.ErrNotExist) {
+			return 0, nil
+		}
+		return 0, err
+	}
+	return byteutil.BytesToUint64BigEndian(h), nil
+}
+
+// PutBlock puts a block into SGDIndexer
+func (sgd *sgdRegistry) PutBlock(ctx context.Context, blk *block.Block) error {
+	var (
+		index    *indexpb.SGDIndex
+		r        *action.Receipt
+		actType  sgdActType
+		contract string
+		ok       bool
+	)
+	if blk == nil {
+		return errors.New("empty block")
```

Comment:
is there an existing error for this? like in /iotex-core/blockchain/block?

Reply:
removed, this check to solve the panic caused by passing nil in the test